### ISSUE TITLE
Custom loss function - MAE

### ISSUE THREAD

**Issue by iradab** [catboost/catboost#2559]
My aim is to replicate MAE loss function of catboost.
According to https://github.com/catboost/catboost/blob/master/catboost/private/libs/algo_helpers/error_functions.h, I see that derivatives are calculated as following:

    double CalcDer(double approx, float target) const override {
        const double val = target - approx;
        if (abs(val) < Delta) return 0;
        return (target - approx > 0) ? Alpha : -(1 - Alpha);
    }
    double CalcDer2(double /*approx*/, float /*target*/) const override {
        return QUANTILE_DER2_AND_DER3;
    }
    double CalcDer3(double /*approx*/, float /*target*/) const override {
        return QUANTILE_DER2_AND_DER3;
    }
Following this tutorial https://github.com/catboost/catboost/blob/master/catboost/tutorials/custom_loss/custom_loss_and_metric_tutorial.ipynb , I write this objective function, which follows the same logics as in source code:

```
class MAEObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)
        result = []
        delta = 10**(-6)
        for i in range(len(targets)):
            diff = targets[i] - approxes[i]
            der1 = 0 if abs(diff) < delta else(0.5 if diff>0 else -0.5)
            der2 = 0
            result.append((der1, der2))
        return result
```

**Issue**: Even though my MAE is converging during training, there is big difference in results of MAEObjective as compared to original MAE. What could be the issue here?

**Comment by andrey-khropov** [comment]
I cannot reproduce the difference.

I've tried the example from the tutorial and plugged in your `MAEObjective` code. And I see identical results.

```py
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

import catboost as cb


print(f'catboost version = {cb.version.VERSION}')


class MAEObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)
        result = []
        delta = 10**(-6)
        for i in range(len(targets)):
            diff = targets[i] - approxes[i]
            der1 = 0 if abs(diff) < delta else(0.5 if diff>0 else -0.5)
            der2 = 0
            result.append((der1, der2))
        return result


X, y = make_regression(random_state=0)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

print('model1')

model1 = cb.CatBoostRegressor(iterations=10, loss_function='MAE', eval_metric='RMSE',
                           learning_rate=0.03, bootstrap_type='Bayesian', boost_from_average=False,
                           leaf_estimation_iterations=1, leaf_estimation_method='Gradient')
model1.fit(X_train, y_train, eval_set=(X_test, y_test))


print('model2')

model2 = cb.CatBoostRegressor(iterations=10, loss_function=MAEObjective(), eval_metric='RMSE',
                           learning_rate=0.03, bootstrap_type='Bayesian', boost_from_average=False,
                           leaf_estimation_iterations=1, leaf_estimation_method='Gradient')
model2.fit(X_train, y_train, eval_set=(X_test, y_test))
```

output:

```
catboost version = 1.2.2
model1
0:	learn: 129.7384858	test: 140.0945654	best: 140.0945654 (0)	total: 59.3ms	remaining: 534ms
1:	learn: 129.7348634	test: 140.0943957	best: 140.0943957 (1)	total: 60.7ms	remaining: 243ms
2:	learn: 129.7316099	test: 140.0938014	best: 140.0938014 (2)	total: 62.1ms	remaining: 145ms
3:	learn: 129.7282234	test: 140.0930445	best: 140.0930445 (3)	total: 63.7ms	remaining: 95.5ms
4:	learn: 129.7255111	test: 140.0934110	best: 140.0930445 (3)	total: 65ms	remaining: 65ms
5:	learn: 129.7221504	test: 140.0934214	best: 140.0930445 (3)	total: 66.3ms	remaining: 44.2ms
6:	learn: 129.7185343	test: 140.0902009	best: 140.0902009 (6)	total: 67.8ms	remaining: 29ms
7:	learn: 129.7154413	test: 140.0894746	best: 140.0894746 (7)	total: 69.6ms	remaining: 17.4ms
8:	learn: 129.7113020	test: 140.0887881	best: 140.0887881 (8)	total: 71.4ms	remaining: 7.94ms
9:	learn: 129.7085913	test: 140.0900800	best: 140.0887881 (8)	total: 73.4ms	remaining: 0us

bestTest = 140.0887881
bestIteration = 8

Shrink model to first 9 iterations.
model2
0:	learn: 129.7384858	test: 140.0945654	best: 140.0945654 (0)	total: 560ms	remaining: 5.04s
1:	learn: 129.7348634	test: 140.0943957	best: 140.0943957 (1)	total: 561ms	remaining: 2.25s
2:	learn: 129.7316099	test: 140.0938014	best: 140.0938014 (2)	total: 563ms	remaining: 1.31s
3:	learn: 129.7282234	test: 140.0930445	best: 140.0930445 (3)	total: 565ms	remaining: 848ms
4:	learn: 129.7255111	test: 140.0934110	best: 140.0930445 (3)	total: 567ms	remaining: 567ms
5:	learn: 129.7221504	test: 140.0934214	best: 140.0930445 (3)	total: 569ms	remaining: 379ms
6:	learn: 129.7185343	test: 140.0902009	best: 140.0902009 (6)	total: 571ms	remaining: 245ms
7:	learn: 129.7154413	test: 140.0894746	best: 140.0894746 (7)	total: 573ms	remaining: 143ms
8:	learn: 129.7113020	test: 140.0887881	best: 140.0887881 (8)	total: 575ms	remaining: 63.8ms
9:	learn: 129.7085913	test: 140.0900800	best: 140.0887881 (8)	total: 576ms	remaining: 0us

bestTest = 140.0887881
bestIteration = 8

Shrink model to first 9 iterations.
```

So please provide more details about how to reproduce.

**Comment by andrey-khropov** [comment]
No info provided for a long time, I'm closing this issue. Feel free to reopen with additional information.
